packages/state-transition/src/cache/epochCache.ts

```diff
@@ -505,16 +572,25 @@ export class EpochCache {
       nextEpochTotalActiveBalanceByIncrement: number;
     }
   ): void {
+    this.previousDecisionRoot = this.currentDecisionRoot;
     this.previousShuffling = this.currentShuffling;
-    this.currentShuffling = this.nextShuffling;
-    const currEpoch = this.currentShuffling.epoch;
-    const nextEpoch = currEpoch + 1;
-
-    this.nextShuffling = computeEpochShuffling(
-      state,
-      epochTransitionCache.nextEpochShufflingActiveValidatorIndices,
-      nextEpoch
-    );
+
+    this.currentDecisionRoot = this.nextDecisionRoot;
+    if (this.nextShuffling) {
+      // was already pulled by the api or another method on EpochCache
+      this.currentShuffling = this.nextShuffling;
+    } else {
+      this.currentShuffling =
+        this.shufflingCache?.buildSync(this.nextEpoch, this.nextDecisionRoot, state, this.nextActiveIndices) ??
+        computeEpochShuffling(state, this.nextActiveIndices, this.nextEpoch);
+    }
+
+    const currentEpoch = this.nextEpoch;
+    this.nextShuffling = null;
+    this.nextDecisionRoot = getShufflingDecisionBlock(state, currentEpoch + 1);
+    this.nextActiveIndices = epochTransitionCache.nextEpochShufflingActiveValidatorIndices;
+    // TODO move this out to beacon node
+    void this.shufflingCache?.build(currentEpoch + 1, this.nextDecisionRoot, state, this.nextActiveIndices);
```

Comment:
if there is no ShufflingCache provided, should compute `nextShuffling` earlier here? then the above code does not need to handle null `nextShuffling` if no ShufflingCache, and hopefully we can remove `MockShufflingCache` because nothing change in case of no ShufflingCache provided

Reply:
I'm not entirely sure why `MockShufflingCache` is needed even now?

epoch caches are created with a `currentShuffling`, which is used in the state transition, and only here in `afterProcessEpoch` is the next `currentShuffling` calculated (L583).

Reply:
> if there is no ShufflingCache provided, should compute nextShuffling earlier here?

Good idea! I'll try that @twoeths

> and hopefully we can remove MockShufflingCache because nothing change in case of no ShufflingCache provided

I'm almost done with all the tests and think i can remove the `MockShufflingCache`.  Last test to check is for `CachedBeaconNode` and i will work through that today.